src/test/perl/path.t

```diff
@@ -290,44 +312,137 @@ rmtree if -d $basetest;
-makefile("$basetest/tgtdir/tgtfile");
-ok(symlink("tgtdir", $dirlink), "directory symlink created");
-ok(symlink("tgtdir/tgtfile", $filelink), "file symlink created");
+=cut
+
+# Test symlink creation
+
+init_exception("symlink tests");
+
+my $target_directory = "tgtdir";
+my $target_file1 = "tgtfile1";
+my $target_file2 = "$basetest/$target_directory/tgtfile2";
+makefile("$basetest/$target_file1");
+makefile($target_file2);
+my %opts;
+is($mc->symlink($target_directory, $dirlink), CHANGED, "directory symlink created");
+ok($mc->is_symlink($dirlink), "$dirlink is a symlink");
+is(readlink($dirlink), $target_directory, "$dirlink has the expected target ($target_directory)");
+is($mc->symlink($target_file2, $filelink), CHANGED, "file symlink created");
+ok($mc->is_symlink($filelink), "$filelink is a symlink");
+is(readlink($filelink), $target_file2, "$filelink has the expected target ($target_file2)");
+is($mc->symlink($target_file2, $filelink), SUCCESS, "file symlink already exists: nothing done");
+ok($mc->is_symlink($filelink), "$filelink is still a symlink");
+is(readlink($filelink), $target_file2, "$filelink target not changed");
+is($mc->symlink($target_file1, $filelink), CHANGED, "file symlink already exists: nothing done");
+ok($mc->is_symlink($filelink), "$filelink is still a symlink (target changed)");
+is(readlink($filelink), $target_file1, "$filelink has the expected target ($target_file1)");
+my $link_status = $mc->symlink($target_file1, $target_file2);
+ok(! defined($link_status), "symlink failed: existing file not replaced by a symlink");
+is($mc->{fail},
+   "*** cannot symlink $target_file2: it is not an existing symlink",
+   "fail attribute set after symlink failure (existing file not replaced by a symlink)");
+ok($mc->file_exists($target_file2) && ! $mc->is_symlink($target_file2), "File $target_file2 has not be replaced by a symlink");
+$opts{force} = 1;
+is($mc->symlink($target_file1, $target_file2, %opts), CHANGED, "existing file replaced by a symlink (force option set)");
+ok($mc->is_symlink($target_file2), "$target_file2 changed to a symlink");
+is(readlink($target_file2), $target_file1, "$target_file2 has the expected target ($target_file1)");
+my $test_directory = "$basetest/$target_directory";
+$link_status = $mc->symlink($target_file1, "$test_directory", %opts);
+ok (! defined($link_status), "directory not replaced by a symlink (force option set)");
+is($mc->{fail},
+   "*** cannot symlink $test_directory: it is not an existing symlink",
+   "fail attribute set after symlink failure (existing directory not replaced by a symlink)");
+ok($mc->directory_exists($test_directory) && ! $mc->is_symlink($test_directory),
+   "Directory $test_directory has not be replaced by a symlink");
+$opts{check} = 1;
+ok(! $mc->symlink("really_really_missing", $brokenlink, %opts), "broken symlink not created (target existence enforced)");
+ok(! -e $brokenlink, "Broken link has not been created");
```

Comment:
this is not a test for a missing broken symlink, only `! -e $fn && ! -l $fn` is. this  one of the reason why broken symlinks are a real mess.

with broken symlink b
```
[stdweird@gast039e tmp]$ perl -e 'print -e "b", "X", -f "b", "X", -l "b", "\n"'
XX1
```